Clappr/Classes/Base/EventDispatcher.swift

```diff
@@ -0,0 +1,123 @@
+    @discardableResult
+    open func once(_ eventName: String, callback: @escaping EventCallback) -> String {
+        return once(eventName, callback: callback, contextObject: self)
+    }
+
+    @discardableResult
+    fileprivate func once(_ eventName: String, callback: @escaping EventCallback, contextObject: EventDispatcher) -> String {
+        let listenId = on(eventName, callback: callback, contextObject: contextObject)
+        onceEventsHashes.append(listenId)
+        return listenId
+    }
+
+    open func off(_ listenId: String) {
+        guard let event = events[listenId] else {
+            Logger.logError("could not find any event with given event listenId: \(listenId)", scope: logIdentifier())
+            return
+        }
+
+        notificationCenter().removeObserver(event.eventHandler, name: NSNotification.Name(rawValue: event.name), object: event.contextObject)
+        events.removeValue(forKey: listenId)
+    }
+
+    open func trigger(_ eventName: String) {
+        trigger(eventName, userInfo: [:])
+    }
+
+    open func trigger(_ eventName: String, userInfo: [AnyHashable: Any]?) {
+        notificationCenter().post(name: Notification.Name(rawValue: eventName), object: self, userInfo: userInfo)
+
+        if type(of: self) != EventDispatcher.self {
+            Logger.logDebug("[\(eventName)] triggered with \(String(describing: userInfo))", scope: logIdentifier())
+        }
+    }
+
+    @discardableResult
+    open func listenTo<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
+        return on(eventName, callback: callback, contextObject: contextObject.getEventDispatcher())
+    }
+
+    @discardableResult
+    open func listenToOnce<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
+        return once(eventName, callback: callback, contextObject: contextObject.getEventDispatcher())
+    }
+
+    open func stopListening() {
+        for (_, event) in events {
+            notificationCenter().removeObserver(event.eventHandler)
+        }
+
+        events.removeAll()
+    }
+
+    open func stopListening(_ listenId: String) {
+        off(listenId)
+    }
+
+    open func getEventDispatcher() -> EventDispatcher {
+        return self
+    }
+
+    fileprivate func createListenId(_ eventName: String, contextObject: EventDispatcher) -> String {
+        let contextObjectHash = ObjectIdentifier(contextObject).hashValue
+        return eventName + String(contextObjectHash) + String(Date().timeIntervalSince1970)
+    }
+
+    fileprivate func notificationCenter() -> NotificationCenter {
+        return NotificationCenter.default
+    }
+
+    fileprivate func logIdentifier() -> String {
+        if let plugin = self as? Plugin {
+            return plugin.pluginName
+        }
+        return "\(type(of: self))"
+    }
+
+    deinit {
+        Logger.logDebug("deinit", scope: NSStringFromClass(type(of: self)))
```

Comment:
on scope: What you think about using the function logIdentifier?

Reply:
Yeah. Good catch. I'm going to do it.